### PR TITLE
Zen 547 save feature file (backend)

### PR DESCRIPTION
**Pull request by daniel112**
**Ticket**: [issue_id](https://jira.simpleviewtools.com/browse/ZEN-547)

## Context

* we need the backend api to support saving files
* File should be saved based on the custom test root path
* frontend implementation is handled on a diff ticket https://jira.simpleviewtools.com/browse/ZEN-533
 
## Goal

* Update the `save` endpoint to allow saving files relative to the custom test root path
* Should throw an error if the given path does not match the test root path

## Updates

* `repos/backend/endpoints/files.js`
  * update to read from the body prop
* `repos/backend/libs/fileSys/testFiles.js`
  * update to save the file at the given location if it's relative to the test root path   

### Setup
* Pull this branch
* run `keg herkin start --config ./repos/example/herkin.config.js --from keg-herkin:zen-547-save-feature-file`
  * this mounts the tests dir of `repos/example` 
* utilize Postman to make a POST call to http://0.0.0.0:5005/files/save
  * [example](https://user-images.githubusercontent.com/3317835/108562912-d6057780-72bd-11eb-9303-0d08f920b06e.png)
* Execute the call with Body:
```JSON
"path" : "some/path"
"content": "some string content" 
``` 
 
### Test 1
* Execute a POST call with the following body:
```JSON
	"path": "/keg/tap/tests/bdd/features/fooBar.feature",
	"content": "new content"
```
* Verify that the call is successful and the file `fooBar.feature` is also added locally at `repos/example/tests/bdd/features`
  * [postman result](https://user-images.githubusercontent.com/3317835/108563661-0568b400-72bf-11eb-8ffb-2b86e3ab5332.png)
  * [file added](https://user-images.githubusercontent.com/3317835/108563700-19acb100-72bf-11eb-805f-5531c55f74d0.png)
* Run the POST call again on an existing path. i.e `/keg/tap/tests/bdd/features/example.feature`
* Ensure that it updates the file with the new content

### Test 2
 * Run the POST call with an invalid test path. (i.e `/Desktop/features/fooBar.feature`)
 * Verify that it fails and returns an error message
   * [image](https://user-images.githubusercontent.com/3317835/108563896-7dcf7500-72bf-11eb-9a8b-210d5a4841de.png)
* **Note:**
  * This `rootPath` is dynamic and is determined by the consumer's herkin config

### CODE REVIEW

File: repos/backend/libs/fileSys/testFiles.js [hunk 1]
```diff
@@ -49,20 +49,31 @@ const getTestFile = async (config, testPath) => {
...
+ * @param {string} fullPath
+ * @param {string} content 
+ */
+const saveTestFile = async (config, fullPath, content) => {
 
-const saveTestFile = async (config, testPath, content) => {
   const { testsRoot } = config.paths
-  const fullPath = path.join(testsRoot, testPath)
+  const inTestRoot = fullPath.startsWith(testsRoot)
+  if (!inTestRoot) throw new Error(`[API - Files] Filepath must have rootPath '${testsRoot}': received '${fullPath}'`)
```
Comment (lancetipton) [inline repos/backend/libs/fileSys/testFiles.js:62]: Typically we don't want to expose paths to the front end for safety reasons. 
Can you update this to say something like:
"File must be saved to the mounted test folder!"

File: repos/backend/libs/fileSys/testFiles.js [hunk 2]
```diff
@@ -49,20 +49,31 @@ const getTestFile = async (config, testPath) => {
...
+ * @param {Object} config 
+ * @param {string} fullPath
+ * @param {string} content 
+ */
+const saveTestFile = async (config, fullPath, content) => {
 
-const saveTestFile = async (config, testPath, content) => {
   const { testsRoot } = config.paths
-  const fullPath = path.join(testsRoot, testPath)
+  const inTestRoot = fullPath.startsWith(testsRoot)
```
Comment (lancetipton) [inline repos/backend/libs/fileSys/testFiles.js:61]: This is a good check, but we should add further validation to this.

We should clean the string by removing special characters, except for '/'
  * Do a google search and check what characters are valid in a file name for Linux systems
  * We definitely want to remove ';' '&&' and ':', but I'm sure there's more

File: repos/backend/endpoints/files.js [hunk 3]
```diff
@@ -10,10 +10,17 @@ const path = require('path')
 
 const saveFile = (app, config) => async (req, res) => {
   try {
-    const file = req.params.file
-    const content = req.params.content
-    const meta = await saveTestFile(config, file)
+    const path = req.body.path
```
Comment (lancetipton) [inline repos/backend/endpoints/files.js:13]: As discussed, we should try not to use 'path' as a variable name. How about 'location' ?

**Review by lancetipton — CHANGES_REQUESTED** [review]
Looks good, but needs a couple quick fixes. 
Thanks

File: repos/backend/libs/fileSys/testFiles.js (hunk 1 above)
Comment (daniel112) [reply repos/backend/libs/fileSys/testFiles.js:62]: updated

File: repos/backend/endpoints/files.js (hunk 3 above)
Comment (daniel112) [reply repos/backend/endpoints/files.js:13]: updated

File: repos/backend/libs/fileSys/testFiles.js (hunk 2 above)
Comment (daniel112) [reply repos/backend/libs/fileSys/testFiles.js:61]: Just wanted to get a full understanding of this edge case.
in what use scenario would this be an issue that the `writeFile` wouldnt catch if it fails?
if a user wants to save a file as `&&.js` for whatever reason shouldn't just let the system decides if that's allowed or not?

Comment (lancetipton) [reply repos/backend/libs/fileSys/testFiles.js:61]: Let's say your on a windows machine. You mount in your test folder

Then you run keg-herkin
Create a new test file, and call it 'awesome:tests.feature'

You write your tests, and everything is working fine. Nothing to worry about. Finish it up and call it a day and go eat a burro.

Tomorrow comes like most days, and you are reviewing your tests on your local file system.

For some reason, you can't seem to open 'awesome:tests.feature' locally. "WTF!", you say out loud.

So you start up keg-herkin, and the test loads fine. 

"WTF!", you say out loud **again**.

So what happened?

A quick google search will most likely tell you => https://www.google.com/search?client=safari&hl=en-us&ei=jro1YMfjEPPi9AOd0Ke4DA&q=windows+invalid+files+names&oq=windows+invalid+files+names&gs_lcp=ChNtb2JpbGUtZ3dzLXdpei1zZXJwEAMyBwgAEEcQsAMyBwgAEEcQsAMyBwgAEEcQsAMyBwgAEEcQsAMyBwgAEEcQsAMyBwgAEEcQsAMyBwgAEEcQsAMyBwgAEEcQsANQAFgAYL4VaAFwAHgAgAEAiAEAkgEAmAEAyAEIwAEB&sclient=mobile-gws-wiz-serp

Then to fix it, you install https://github.com/sindresorhus/filenamify

Later you see Lance, and you thank him for being awesome.

Comment (daniel112) [reply repos/backend/libs/fileSys/testFiles.js:62]: best user story ever.
updated

**Review by lancetipton — CHANGES_REQUESTED** [review]
Code looks good. There's just some formatting stuff. Can you just run the yarn format command, should fix it. Thanks

File: repos/backend/endpoints/files.js [hunk 4]
```diff
@@ -10,10 +10,17 @@ const path = require('path')
 
 const saveFile = (app, config) => async (req, res) => {
   try {
-    const file = req.params.file
-    const content = req.params.content
-    const meta = await saveTestFile(config, file)
+    const location = req.body.path
+    if (!location)
+    return apiErr(
```
[the remark this answers is not in the document]
Comment (daniel112) [reply repos/backend/endpoints/files.js:15]: updated with the indent

**Review by lancetipton — APPROVED** [review]
Looks good. All tests pass

**Pull request merged**